File: src/application/smart-day.ts
import type { Event, Plan } from '../domain/models'

export type TimedKind = 'event' | 'plan'

export interface TimedBlock {
  kind: TimedKind
  id: string
  title: string
  startAt: string
  endAt: string
  color: string
}

export interface SmartDaySnapshot {
  now: string
  status: 'busy' | 'free'
  label: string
  current: TimedBlock | null
  next: TimedBlock | null
  elapsedMs: number | null
  remainingMs: number | null
  overlaps: TimedBlock[]
}

function toBlock(kind: TimedKind, item: Event | Plan): TimedBlock {
  return {
    kind,
    id: item.id,
    title: item.title,
    startAt: item.startAt,
    endAt: item.endAt,
    color: item.color,
  }
}

function isActiveTimedEvent(event: Event): boolean {
  return event.deletedAt === undefined && !event.allDay
}

function isActivePlan(plan: Plan): boolean {
  return plan.deletedAt === undefined
}

/** Build Smart Day view from timed Events + Plans. All-day events and tasks are ignored. */
export function buildSmartDaySnapshot(
  events: Event[],
  plans: Plan[],
  nowIso: string,
): SmartDaySnapshot {
  const nowMs = Date.parse(nowIso)
  const blocks: TimedBlock[] = [
    ...events.filter(isActiveTimedEvent).map((event) => toBlock('event', event)),
    ...plans.filter(isActivePlan).map((plan) => toBlock('plan', plan)),
  ].sort((left, right) => left.startAt.localeCompare(right.startAt) || left.endAt.localeCompare(right.endAt))

  const overlaps = blocks.filter((block) => {
    const start = Date.parse(block.startAt)
    const end = Date.parse(block.endAt)
    return start <= nowMs && nowMs < end
  })

  const current = overlaps[0] ?? null
  const next =
    blocks
      .filter((block) => Date.parse(block.startAt) > nowMs)
      .sort((left, right) => left.startAt.localeCompare(right.startAt))[0] ?? null

  if (current === null) {
    return {
      now: nowIso,
      status: 'free',
      label: 'Сейчас свободно',
      current: null,
      next,
      elapsedMs: null,
      remainingMs: null,
      overlaps: [],
    }
  }

  const startMs = Date.parse(current.startAt)
  const endMs = Date.parse(current.endAt)
  return {
    now: nowIso,
    status: 'busy',
    label: current.title,
    current,
    next,
    elapsedMs: Math.max(0, nowMs - startMs),
    remainingMs: Math.max(0, endMs - nowMs),
    overlaps,
  }
}

export function formatDuration(ms: number): string {
  const totalMinutes = Math.floor(ms / 60_000)
  const hours = Math.floor(totalMinutes / 60)
  const minutes = totalMinutes % 60
  if (hours <= 0) return `${minutes} мин`
  if (minutes === 0) return `${hours} ч`
  return `${hours} ч ${minutes} мин`
}

File: backend/app/application/smart_day.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict


def to_camel(name: str) -> str:
    parts = name.split("_")
    return parts[0] + "".join(part.title() for part in parts[1:])


class CamelModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class TimedBlock(CamelModel):
    kind: Literal["event", "plan"]
    id: str
    title: str
    start_at: str
    end_at: str
    color: str


class SmartDaySnapshot(CamelModel):
    now: str
    status: Literal["busy", "free"]
    label: str
    current: TimedBlock | None
    next: TimedBlock | None
    elapsed_ms: int | None
    remaining_ms: int | None
    overlaps: list[TimedBlock]


def _parse(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00")).replace(tzinfo=None)


def build_smart_day_snapshot(
    events: list[dict[str, Any]],
    plans: list[dict[str, Any]],
    now_iso: str,
) -> SmartDaySnapshot:
    now = _parse(now_iso)
    blocks: list[TimedBlock] = []

    for event in events:
        if event.get("deletedAt") or event.get("allDay"):
            continue
        blocks.append(
            TimedBlock(
                kind="event",
                id=event["id"],
                title=event["title"],
                start_at=event["startAt"],
                end_at=event["endAt"],
                color=event["color"],
            )
        )

    for plan in plans:
        if plan.get("deletedAt"):
            continue
        blocks.append(
            TimedBlock(
                kind="plan",
                id=plan["id"],
                title=plan["title"],
                start_at=plan["startAt"],
                end_at=plan["endAt"],
                color=plan["color"],
            )
        )

    blocks.sort(key=lambda item: (item.start_at, item.end_at))
    overlaps = [
        block
        for block in blocks
        if _parse(block.start_at) <= now < _parse(block.end_at)
    ]
    current = overlaps[0] if overlaps else None
    future = [block for block in blocks if _parse(block.start_at) > now]
    next_block = sorted(future, key=lambda item: item.start_at)[0] if future else None

    if current is None:
        return SmartDaySnapshot(
            now=now_iso,
            status="free",
            label="Сейчас свободно",
            current=None,
            next=next_block,
            elapsed_ms=None,
            remaining_ms=None,
            overlaps=[],
        )

    start = _parse(current.start_at)
    end = _parse(current.end_at)
    elapsed = max(0, int((now - start).total_seconds() * 1000))
    remaining = max(0, int((end - now).total_seconds() * 1000))
    return SmartDaySnapshot(
        now=now_iso,
        status="busy",
        label=current.title,
        current=current,
        next=next_block,
        elapsed_ms=elapsed,
        remaining_ms=remaining,
        overlaps=overlaps,
    )


def plan_progress(tasks: list[dict[str, Any]]) -> dict[str, int]:
    active = [task for task in tasks if not task.get("deletedAt")]
    completed = sum(1 for task in active if task.get("completed"))
    total = len(active)
    percentage = 0 if total == 0 else round((completed / total) * 100)
    return {"completed": completed, "total": total, "percentage": percentage}
